packages/core/contracts/cross-chain-oracle/chain-adapters/Polygon_ChildMessenger.sol

```diff
@@ -0,0 +1,86 @@
+
+    /**
+     * @notice Construct the Polygon_ChildMessenger contract.
+     * @param _fxChild Polygon system contract deployed on Mainnet, required to construct new FxBaseRootTunnel
+     * that can send messages via native Polygon data tunnel.
+     */
+    constructor(address _fxChild) FxBaseChildTunnel(_fxChild) {}
+
+    /**
+     * @notice Set OracleSpoke address, which is the only address that can call `sendMessageToParent`.
+     * @dev Can only reset this address once.
+     * @param _oracleSpoke address of the new oracle spoke, deployed on this network.
+     */
+    function setOracleSpoke(address _oracleSpoke) public nonReentrant() {
+        require(oracleSpoke == address(0x0), "OracleSpoke already set");
+        oracleSpoke = _oracleSpoke;
+        emit SetOracleSpoke(oracleSpoke);
+    }
+
+    /**
+     * @notice Set OracleHub address, which is always the target address for messages sent from this network to
+     * the parent network.
+     * @dev Can only reset this address once.
+     * @param _oracleHub address of the new oracle hub, deployed on the parent network.
+     */
+    function setOracleHub(address _oracleHub) public nonReentrant() {
+        require(oracleHub == address(0x0), "OracleHub already set");
+        oracleHub = _oracleHub;
+        emit SetOracleHub(oracleHub);
+    }
+
+    /**
+     * @notice Sends a message to the OracleSpoke via the parent messenger and the canonical message bridge.
+     * @dev The caller must be the OracleSpoke on child network. No other contract is permissioned to call this
+     * function.
+     * @dev The L1 target, the parent messenger, must implement processMessageFromChild to consume the message.
+     * @param data data message sent to the L1 messenger. Should be an encoded function call or packed data.
+     */
+    function sendMessageToParent(bytes memory data) public override nonReentrant() {
+        require(msg.sender == oracleSpoke, "Only callable by oracleSpoke");
+        _sendMessageToRoot(abi.encode(data, oracleHub));
+        emit MessageSentToParent(data, oracleHub);
+    }
+
+    /**
+     * @notice Process a received message from the parent messenger via the canonical message bridge.
+     * @dev The data will be received automatically from the state receiver when the state is synced between Ethereum
+     * and Polygon. This will revert if the Root chain sender is not the `fxRootTunnel` contract.
+     * @dev This call will revert if `setFxRoot` has not been called. Fx Child should be set to Polygon_ParentMessenger.
+     * @param sender The sender of `data` from the Root chain.
+     * @param data ABI encoded params with which to call function on OracleHub or GovernorHub.
+     */
+    function _processMessageFromRoot(
+        uint256, /* stateId */
+        address sender,
+        bytes memory data
+    ) internal override validateSender(sender) nonReentrant() {
```

Comment:
where is this `validateSender` coming from? how does that work? a comment might be needed as I cant find this in the PR anywhere

Reply:
Its in `external/polygon/FxBaseChildTunnel`.  I edited the comments.